utxoscanner.go

```diff
@@ -0,0 +1,517 @@
+	// in the event we experience a critical rescan error.
+	spillRequests := func() []*GetUtxoRequest {
+		requestList := make([]*GetUtxoRequest, len(requests))
+		for _, req := range requests {
+			requestList = append(requestList, req)
+		}
+		return requestList
+	}
+
+	// Take the request with the lowest block height so we can begin the scan
+	// from there.
+	req := s.peek()
+
+	// Check if we were interrupted while waiting.
+	if req == nil {
+		return nil, nil
+	}
+
+	// While scanning through the blockchain, take note of the transactions that
+	// create the outpoints. If the outpoint isn't spent then return this
+	// transaction.
+	initialTx := make(map[wire.OutPoint]*SpendReport)
+	outpoints := make(map[wire.OutPoint][]byte)
+	var filterEntries [][]byte
+
+	bestStamp, err := s.cfg.BestSnapshot()
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		startHeight = req.StartHeight
+		endHeight   = uint32(bestStamp.Height)
+	)
+
+scanToEnd:
+	// Scan forward through the blockchain and look for any transactions that
+	// might spend the given UTXOs.
+	for height := startHeight; height <= endHeight; height++ {
+		// Check to see if the utxoscanner has been signaled to exit.
+		select {
+		case <-s.quit:
+			return spillRequests(), ErrShuttingDown
+		default:
+		}
+
+		// If there are any new requests that can safely be added to this batch,
+		// then try and fetch them.
+		reqs := s.dequeueAtHeight(height)
+		for _, req := range reqs {
+			log.Debugf("Adding outpoint=%s height=%d to watchlist",
+				req.OutPoint, req.StartHeight)
+
+			entry := builder.OutPointToFilterEntry(*req.OutPoint)
+			filterEntries = append(filterEntries, entry)
+
+			outpoints[*req.OutPoint] = entry
+			requests[*req.OutPoint] = req
+		}
+
+		hash, err := s.cfg.GetBlockHash(int64(height))
+		if err != nil {
+			return spillRequests(), err
+		}
+
+		// If an outpoint is created in this block, then fetch it regardless.
+		// Otherwise check to see if the filter matches any of our watched
+		// outpoints.
+		fetch := len(reqs) > 0
+		if !fetch {
+			options := rescanOptions{
+				watchList: filterEntries,
+			}
+
+			match, err := s.cfg.BlockFilterMatches(&options, hash)
+			if err != nil {
+				return spillRequests(), err
+			}
+
+			// If still no match is found, we have no reason to
+			// fetch this block, and can continue to next height.
+			if !match {
+				continue
+			}
+		}
+
+		log.Debugf("Fetching block height=%d hash=%s", height, hash)
+
+		block, err := s.cfg.GetBlock(*hash)
+		if err != nil {
+			return spillRequests(), err
+		}
+
+		// Check to see if the utxoscanner has been signaled to exit.
+		select {
+		case <-s.quit:
+			return spillRequests(), ErrShuttingDown
+		default:
+		}
+
+		log.Debugf("Got block height=%d hash=%s", height, hash)
+
+		// Grab the txns that created the outputs requested at this
+		// block height.
+		txnsInBlock := findInitialTransactions(block.MsgBlock(), reqs)
```

Comment:
Why is is necessary that we even locate the initial set of transactions? To ensure that the output was actually even _created_ in the history of the chain?

Reply:
Yep, in the current impl we walk back and return the still-unspent txn if no spend was detected.